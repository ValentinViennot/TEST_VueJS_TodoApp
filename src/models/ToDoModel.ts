export default interface ToDoModel {
  text: string;
  done: boolean;
  id: number;
}
