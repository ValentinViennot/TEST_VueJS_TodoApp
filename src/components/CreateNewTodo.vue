<template>
  <div class="addContainer">
      <input class="addInput" type="text" v-model="newTodo.text" @keyup.enter="handleAdd" />
      <button class="addButton" @click="handleAdd">Add</button>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import ToDoModel from '../models/ToDoModel';

@Component
export default class CreateNewTodo extends Vue {
  newTodo: ToDoModel = CreateNewTodo.getEmptyTodo();

  handleAdd() {
    if (this.isTodoEmpty()) return;
    this.$emit('new-todo', this.newTodo);
    this.newTodo = CreateNewTodo.getEmptyTodo();
  }

  private isTodoEmpty() {
    return this.newTodo.text.length === 0;
  }

  static getEmptyTodo() {
    return {
      text: '',
      done: false,
      id: -1
    };
  }
}
</script>

<style scoped>
</style>
