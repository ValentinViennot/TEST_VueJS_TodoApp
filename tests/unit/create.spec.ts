import { shallowMount, Wrapper } from '@vue/test-utils';
import CreateNewTodo from '../../src/components/CreateNewTodo.vue';
import ToDoModel from '../../src/models/ToDoModel';

describe('CreateNewTodo.vue component', () => {
  let wrapper: Wrapper<CreateNewTodo>;
  const emptyTodo: ToDoModel = {
    text: '',
    done: false,
    id: -1
  };

  beforeAll(() => {
    wrapper = shallowMount(CreateNewTodo);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when click on "Add"', () => {
    let emitted: any;

    describe('if text field is empty', () => {
      beforeAll(() => {
        wrapper = shallowMount(CreateNewTodo);
        wrapper.find('button').element.click();
        emitted = wrapper.emitted();
      });

      it('should NOT emit a "new-todo" event', () => {
        expect(emitted['new-todo']).toBeFalsy();
      });
    });

    describe('if text field is not empty', () => {
      const dummyText = 'Hello World';
      const dummyTodo: ToDoModel = { ...emptyTodo, text: dummyText };
      beforeAll(() => {
        wrapper = shallowMount(CreateNewTodo);
        wrapper.find('input').setValue(dummyText);
        wrapper.find('input').trigger('change');
        wrapper.find('button').element.click();
        emitted = wrapper.emitted();
      });

      it('should emit a "new-todo" event', () => {
        expect(emitted['new-todo']).toBeTruthy();
      });

      it('should should pass newTodo data through the event', () => {
        expect(emitted['new-todo'][0]).toEqual([dummyTodo]);
      });

      it('should empty the input field', () => {
        expect(wrapper.vm.$data.newTodo.text).toEqual('');
      });
    });
  });

  describe('when hitting Enter on Input field', () => {
    let emitted: any;

    describe('if text field is not empty', () => {
      beforeAll(() => {
        wrapper = shallowMount(CreateNewTodo);
        wrapper.find('input').setValue('dummyText');
        wrapper.find('input').trigger('keyup.enter');
        emitted = wrapper.emitted();
      });

      it('should emit a "new-todo" event', () => {
        expect(emitted['new-todo']).toBeTruthy();
      });
    });
  });
});
