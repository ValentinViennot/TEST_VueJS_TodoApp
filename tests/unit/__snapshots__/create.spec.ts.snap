// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewTodo.vue component should render without crashing 1`] = `<div class="addContainer"><input type="text" class="addInput"> <button class="addButton">Add</button></div>`;
